### PR TITLE
Added article to guide Patterns and Practices for SPFx Development

### PR DESCRIPTION
**Pull request by Dipesh-Bhanani**
## Category

- [ ] Content fix
- [x] New article

## Related issues

- none

## What's in this Pull Request?

Added article to guide Patterns and Practices for SPFx Development. It helps to write beautiful and consistent code for SPFx solutions by implementing best practices.

### CODE REVIEW

**Comment by VesaJuvonen** [comment]
Docs Build status updates of commit _[8b881a6](https://github.com/Dipesh-Bhanani/sp-dev-docs/commits/8b881a6191eab05ba0db27b7bfb87ea5ed485d29)_: 

### :warning: Validation status: warnings


File | Status | Preview URL | Details
---- | ------ | ----------- | -------
[docs/spfx/spfx-patterns-and-practices.md](https://github.com/Dipesh-Bhanani/sp-dev-docs/blob/main/docs/spfx/spfx-patterns-and-practices.md) | :warning:Warning | [View](https://review.docs.microsoft.com/en-us/sharepoint/dev/spfx/spfx-patterns-and-practices?branch=pr-en-us-8358) | [Details](#user-content-e0b2de85b87f9a2c68415dd19041ec2b-c6e71f21)

<a id="e0b2de85b87f9a2c68415dd19041ec2b-c6e71f21"></a>
### [docs/spfx/spfx-patterns-and-practices.md](https://github.com/Dipesh-Bhanani/sp-dev-docs/blob/main/docs/spfx/spfx-patterns-and-practices.md)
 - **Line 224, Column 68**: **[Warning: hard-coded-locale - [See documentation](https://review.docs.microsoft.com/help/contribute/validation-ref/hard-coded-locale?branch=main)]** `````Link 'https://docs.microsoft.com/en-us/javascript/api/sp-core-library/servicescope?view=sp-typescript-latest' contains locale code 'en-us'. For localizability, remove 'en-us' from links to most Microsoft sites.`````
 - **Line 224, Column 68**: **[Suggestion: docs-link-absolute - [See documentation](https://review.docs.microsoft.com/help/contribute/validation-ref/docs-link-absolute?branch=main)]** `````Absolute link 'https://docs.microsoft.com/en-us/javascript/api/sp-core-library/servicescope?view=sp-typescript-latest' will be broken in isolated environments. Replace with a relative link.`````
 - **Line 224, Column 68**: **[Suggestion: preserve-view-not-set - [See documentation](https://review.docs.microsoft.com/help/contribute/validation-ref/preserve-view-not-set?branch=main)]** `````You've pinned this link to a specific version of content with the view parameter. It's recommended not to pin a version unless that version is A) not the default view and B) the context is about that version specifically. To proceed with pinning a version add the &preserve-view=true to the URL.  Otherwise, remove the view parameter. URL: https://docs.microsoft.com/en-us/javascript/api/sp-core-library/servicescope?view=sp-typescript-latest`````

For more details, please refer to the [build report](https://buildapi.docs.microsoft.com/Output/PullRequest/8c4da01d-e430-7fe7-0bea-2c4100cce9e8/202208090748421370-8358/BuildReport?accessString=ef9df1814389ed10006f11cee9fff7be7868273b9146ae0dfc58c677a161004e).

If you see build warnings/errors with permission issues, it might be due to single sign-on (SSO) enabled on Microsoft's GitHub organizations. Please follow instructions [here](https://review.docs.microsoft.com/en-us/help/onboard/sso-troubleshooting?branch=main) to re-authorize your GitHub account to Docs Build.

**Note:** Broken links written as relative paths are included in the above build report. For broken links written as absolute paths or external URLs, see the [broken link report](https://ops.microsoft.com/#/repos/8c4da01d-e430-7fe7-0bea-2c4100cce9e8?tabName=brokenlinks).

**Note:** Your PR may contain errors or warnings unrelated to the files you changed. This happens when external dependencies like GitHub alias, Microsoft alias, cross repo links are updated. Please use these [instructions](https://review.docs.microsoft.com/en-us/help/onboard/faq?branch=main&tabs=pr-validation-error-solution#q-why-do-i-get-pr-validation-result-with-cryptic-error-or-warning-messages-not-related-to-the-changes-i-make-in-my-working-branch) to resolve them.

For any questions, please:<ul><li>Try searching the docs.microsoft.com <a href="https://review.docs.microsoft.com/en-us/help/?branch=main">contributor guides</a></li><li>Post your question in the <a href="https://teams.microsoft.com/l/channel/19%3a7ecffca1166a4a3986fed528cf0870ee%40thread.skype/General?groupId=de9ddba4-2574-4830-87ed-41668c07a1ca&amp;tenantId=72f98bf-86f1-41af-91ab-2d7cd011db47">Docs support channel</a></li></ul>

**Comment by VesaJuvonen** [comment]
Docs Build status updates of commit _[e730b67](https://github.com/Dipesh-Bhanani/sp-dev-docs/commits/e730b678842a087227d192bdac2128f8f0f4ab4b)_: 

### :white_check_mark: Validation status: passed


File | Status | Preview URL | Details
---- | ------ | ----------- | -------
[docs/spfx/spfx-patterns-and-practices.md](https://github.com/Dipesh-Bhanani/sp-dev-docs/blob/main/docs/spfx/spfx-patterns-and-practices.md) | :white_check_mark:Succeeded | [View](https://review.docs.microsoft.com/en-us/sharepoint/dev/spfx/spfx-patterns-and-practices?branch=pr-en-us-8358) |

For more details, please refer to the [build report](https://buildapi.docs.microsoft.com/Output/PullRequest/8c4da01d-e430-7fe7-0bea-2c4100cce9e8/202208090807169623-8358/BuildReport?accessString=96cfacd0d588eafe74ea17cad15b7d5d777ef30649601a92ee0d4a7a25025603).

**Note:** Broken links written as relative paths are included in the above build report. For broken links written as absolute paths or external URLs, see the [broken link report](https://ops.microsoft.com/#/repos/8c4da01d-e430-7fe7-0bea-2c4100cce9e8?tabName=brokenlinks).

For any questions, please:<ul><li>Try searching the docs.microsoft.com <a href="https://review.docs.microsoft.com/en-us/help/?branch=main">contributor guides</a></li><li>Post your question in the <a href="https://teams.microsoft.com/l/channel/19%3a7ecffca1166a4a3986fed528cf0870ee%40thread.skype/General?groupId=de9ddba4-2574-4830-87ed-41668c07a1ca&amp;tenantId=72f98bf-86f1-41af-91ab-2d7cd011db47">Docs support channel</a></li></ul>

File: docs/spfx/spfx-patterns-and-practices.md [hunk 1]
```diff
@@ -0,0 +1,324 @@
...
+* These UI components should only contain rendering logic. Do not add any business logic or data access code in these components.
+* You should not add multiple function components/classes to single file.
+* Consider keeping the code minimal in render method.
+* Consider handling exception properly. Log the exception details with logger components and present generic informative message for end-user on screen. 
+* Do not hardcode strings including generic messages. Use string labels defined in language files (en-us.js) located at src > webparts > myWebPart > loc
+* Do not store large objects in session/local storages
+* Consider importing leaf level components while importing third party libraries. For example, if you want to import PrimayButton from Office Fabric library, use import { PrimaryButton } from "@fluentui/react/lib/Button"; rather than import { PrimaryButton } from "@fluentui/react";. This will help you to keep your package size compact. This can be very useful in scenarios where you would want to build home page components which are required to be performance effective at first load.
```
Comment (andrewconnell) [inline docs/spfx/spfx-patterns-and-practices.md:274]: Use inline code markdown formatting.

**Comment by andrewconnell** [comment]
I'm unclear as to the point of this. Lots of opinions and guidance... is this coming from MS or just your opinions? Where is this referenced/linked from?

As this is a new article, defer to MS for merging.

**Comment by Dipesh-Bhanani** [comment]
@andrewconnell Those are my own opinions based on experience.

**Comment by VesaJuvonen** [comment]
Docs Build status updates of commit _[21bf3e3](https://github.com/Dipesh-Bhanani/sp-dev-docs/commits/21bf3e3079192c31301b1a2fbc2987ca212aa606)_: 

### :white_check_mark: Validation status: passed


File | Status | Preview URL | Details
---- | ------ | ----------- | -------
[docs/spfx/spfx-patterns-and-practices.md](https://github.com/Dipesh-Bhanani/sp-dev-docs/blob/main/docs/spfx/spfx-patterns-and-practices.md) | :white_check_mark:Succeeded | [View](https://review.docs.microsoft.com/en-us/sharepoint/dev/spfx/spfx-patterns-and-practices?branch=pr-en-us-8358) |

For more details, please refer to the [build report](https://buildapi.docs.microsoft.com/Output/PullRequest/8c4da01d-e430-7fe7-0bea-2c4100cce9e8/202208091028196625-8358/BuildReport?accessString=2928a02483a45fe8e96515b340668e5dd1df7aba1f5db5626c179dca1c626080).

**Note:** Broken links written as relative paths are included in the above build report. For broken links written as absolute paths or external URLs, see the [broken link report](https://ops.microsoft.com/#/repos/8c4da01d-e430-7fe7-0bea-2c4100cce9e8?tabName=brokenlinks).

For any questions, please:<ul><li>Try searching the docs.microsoft.com <a href="https://review.docs.microsoft.com/en-us/help/?branch=main">contributor guides</a></li><li>Post your question in the <a href="https://teams.microsoft.com/l/channel/19%3a7ecffca1166a4a3986fed528cf0870ee%40thread.skype/General?groupId=de9ddba4-2574-4830-87ed-41668c07a1ca&amp;tenantId=72f98bf-86f1-41af-91ab-2d7cd011db47">Docs support channel</a></li></ul>

File: docs/spfx/spfx-patterns-and-practices.md (hunk 1 above)
Comment (Dipesh-Bhanani) [reply docs/spfx/spfx-patterns-and-practices.md]: Committed changes with inline code formatting. Thanks.

**Comment by andrewconnell** [comment]
Thanks for clarifying... while the contribution is appreciated. If you take this to the logical conclusion, there could be many different opinions from individuals on different coding styles... so should they all be backed in this page or multiple pages. 

The official SPFx docs are not intended for individual opinions. Rather they're for documentation or guidance coming from Microsoft. I'd suggest you post this to your own personal blog or other community sites.

**Comment by Dipesh-Bhanani** [comment]
> Thanks for clarifying... while the contribution is appreciated, if you take this to the logical conclusion, there could be many different opinions from individuals on different coding styles... so should they all be backed in this page or multiple pages.
> 
> The official SPFx docs are not intended for individual opinions. Rather they're for documentation or guidance coming from Microsoft. I'd suggest you post this to your own personal blog or other community sites.

Thanks @andrewconnell. I expected that coming. I was actually awaiting this kind of guidance from MS only but I couldn't find in sp-dev-docs so thought to contribute here and make it reach broader audience.

**Pull request closed without merging**